sql/api/src/main/scala/org/apache/spark/sql/internal/types/AbstractStringType.scala

```diff
@@ -51,3 +51,14 @@ case object StringTypeBinaryLcase extends AbstractStringType {
 case object StringTypeAnyCollation extends AbstractStringType {
   override private[sql] def acceptsType(other: DataType): Boolean = other.isInstanceOf[StringType]
 }
+
+/**
+ * Use StringTypeNonCSAICollation for expressions supporting all possible collation types
+ * except CS_AI collation types.
+ */
+case object StringTypeNonCSAICollation extends AbstractStringType {
+  override private[sql] def acceptsType(other: DataType): Boolean =
+    other.isInstanceOf[StringType] &&
+      (!other.asInstanceOf[StringType].typeName.contains("_AI") ||
+      other.asInstanceOf[StringType].typeName.contains("_CI"))
```

Comment:
Why do you use the complex expression? From the PR description and function name, you need to exclude only `CS_AI`.

Reply:
Yes, but this also includes the case that the collator has only `_AI` suffix and also we want to allow `CI_AI` collators for these expressions as they are behaving correctly. Also, the modifier may be written like `AI_CS` or just `_AI` like in the PR description example.

Reply:
My thoughts here are that it might not be a nice way to check for sensitivity based of a collation name (?). Perhaps it would be more cleaner to extend the `StringType` object with `isNonCSAICollation` method (like we did with `isUTF8BinaryCollation` and `isUTF8LcaseCollation`.